subsys/net/lib/ftp_client/src/ftp_client.c

```diff
@@ -0,0 +1,735 @@
+	if (strstr(message, code_str)) {
+		ret = success_code;
+	}
+
+	return ret;
+}
+
+static int establish_data_channel(const char *pasv_msg)
+{
+	int ret;
+	char tmp[16];
+	char *tmp1, *tmp2;
+	u16_t data_port;
+	int data_sock;
+
+	/* Parse Server port from passive message
+	 * e.g. "227 Entering Passive Mode (90,130,70,73,86,111)"
+	 */
+	tmp1 = strstr(pasv_msg, "(");
+	tmp1++;
+	tmp2 = strstr(tmp1, ",");
+	tmp2++;
+	tmp1 = strstr(tmp2, ",");
+	tmp1++;
+	tmp2 = strstr(tmp1, ",");
+	tmp2++;
+	tmp1 = strstr(tmp2, ",");
+	tmp1++;
+	tmp2 = strstr(tmp1, ",");
+	memset(tmp, 0x00, 16);
+	strncpy(tmp, (const char *)tmp1, (size_t)(tmp2 - tmp1));
+	data_port = atoi(tmp) << 8;
+	tmp2++;
+	tmp1 = strstr(tmp2, ")");
+	memset(tmp, 0x00, 16);
+	strncpy(tmp, (const char *)tmp2, (size_t)(tmp1 - tmp2));
+	data_port += atoi(tmp);
+
+	/* Establish the second connect for data */
+	if (client.sec_tag <= 0) {
+		data_sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	} else {
+		data_sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	}
+	if (data_sock < 0) {
+		LOG_ERR("socket(data) failed: %d", -errno);
+		ret = -errno;
+	}
+
+	if (client.sec_tag > 0) {
+		sec_tag_t sec_tag_list[1] = { client.sec_tag };
+
+		ret = setsockopt(data_sock, SOL_TLS, TLS_SEC_TAG_LIST,
+				sec_tag_list, sizeof(sec_tag_t));
+		if (ret) {
+			LOG_ERR("set tag list failed: %d", -errno);
+			close(data_sock);
+			return -errno;
+		}
+
+	}
+	client.remote.sin_port = htons(data_port);
+	ret = connect(data_sock, (struct sockaddr *)&client.remote,
+		 sizeof(struct sockaddr_in));
+	if (ret < 0) {
+		LOG_ERR("connect(data) failed: %d", -errno);
+		close(data_sock);
+		return -errno;
+	}
+
+	return data_sock;
+}
+
+/**@brief Send FTP message via socket
+ */
+static int do_ftp_send_ctrl(const u8_t *message, int length)
+{
+	int ret;
+	u32_t offset = 0;
+
+	LOG_HEXDUMP_DBG(message, length, "TXC");
+
+	while (offset < length) {
+		ret = send(client.sock, message + offset, length - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send(ctrl) failed: %d", -errno);
+			ret = -errno;
+			break;
+		}
+		offset += ret;
+		ret = 0;
+	}
+
+	LOG_DBG("CMD sent");
+	return ret;
+}
+
+/**@brief Send FTP data via socket
+ */
+static int do_ftp_send_data(const char *pasv_msg, u8_t *message, u16_t length)
+{
+	int data_sock;
+	int ret;
+	u32_t offset = 0;
+
+	/* Establish data channel */
+	ret = establish_data_channel(pasv_msg);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	LOG_HEXDUMP_DBG(message, length, "TXD");
+
+	data_sock = ret;
+	while (offset < length) {
+		ret = send(data_sock, message + offset, length - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send(data) failed: %d", -errno);
+			ret = -errno;
+			break;
+		}
+		offset += ret;
+		ret = 0;
+	}
+
+	LOG_DBG("DATA sent");
+	close(data_sock);
+	return ret;
+}
+
+/**@brief Receive FTP message from socket
+ */
+static int do_ftp_recv_ctrl(bool post_result, int success_code)
+{
+	int ret;
+	struct pollfd fds[1];
+
+	memset(ctrl_buf, 0x00, sizeof(ctrl_buf));
```

Comment:
A follow up on https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/2178#discussion_r426387172

> After calling do_ftp_recv_ctrl(), content of ctrl_buf is checked. In case the caller not check the return code (e.g. recv failed scenario), make sure the buffer is empty.

I'm sorry but I don't buy that. In case of reception error (which is indicated clearly by this function returning negative error code), no assumption shall be made about the content of the receive buffer. We shall not waste CPU cycles as a precaution of some improper use of this function, the user is responsible to verify return code.

Reply:
Understood, will remove the redundant logic